container_push: added attribute 'tag_file'

excavador opened the pull request:
I need ability to specify tag of image for push as content of file.
I added `tag_file` attribute to `container_push` rule.

xingao267 commented on container/push.bzl:
```diff
@@ -167,6 +171,10 @@ container_push = rule(
         ),
+        "tag_file": attr.label(
+            allow_single_file = True,
+            doc = "(options) The label of the file with tag value. Overrides 'tag'.",
```
(optional)

xingao267 commented on container/push.bzl:
```diff
@@ -99,6 +99,10 @@ def _impl(ctx):
         "--stamp-info-file=%s" % _get_runfile_path(ctx, f)
         for f in stamp_inputs
     ]
+    runfiles_tag_file = []
```
Move this new section to be after line 90 so it's more clear how tag is set.

excavador commented:
Will do, thank you!

excavador commented:
> One that is similar to https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/BUILD#L262 and https://github.com/bazelbuild/rules_docker/blob/master/testing/e2e.sh#L390

Done

> And another file_test (example: https://github.com/GoogleContainerTools/base-images-docker/blob/master/tests/package_managers/BUILD#L32) that validates that the generated executable file has the desired content (e.g. the tag that reads from a file)

Should I add this test to rules_docker or to https://github.com/GoogleContainerTools/base-images-docker/?

xingao267 commented:
In rules_docker as well. Probably in file
https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/BUILD
as well. Similar to other file_test.

On Fri, Feb 22, 2019 at 7:23 AM Oleg Tsarev <notifications@github.com>
wrote:

> One that is similar to
> https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/BUILD#L262
> and
> https://github.com/bazelbuild/rules_docker/blob/master/testing/e2e.sh#L390
>
> Done
>
> And another file_test (example:
> https://github.com/GoogleContainerTools/base-images-docker/blob/master/tests/package_managers/BUILD#L32)
> that validates that the generated executable file has the desired content
> (e.g. the tag that reads from a file)
>
> Should I add this test to rules_docker or to
> https://github.com/GoogleContainerTools/base-images-docker/?
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_docker/pull/691#issuecomment-466379212>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKBGUiDdeUYFjE2AKrUF2vIuxaCEYCJJks5vP-FKgaJpZM4a8Yaj>
> .
>

excavador commented:
> In rules_docker as well. Probably in file https://github.com/bazelbuild/rules_docker/blob/master/tests/docker/BUILD as well. Similar to other file_test.

I see. Done.

excavador commented:
@xingao267 fixed PR comments

excavador commented:
rebased to latest master

excavador commented:
@xingao267 everything fixed, tests added, PR rebased to latest master.

xingao267 commented on tests/docker/BUILD:
```diff
@@ -285,6 +304,12 @@ file_test(
 )
 
+file_test(
+    name = "test_push_tag_file_digest_output",
```
Don't check the digest output, check the default output executable that it contains the right tag.

xingao267 commented on tests/docker/BUILD:
```diff
@@ -269,6 +269,25 @@ container_push(
+    tags = ["manual"],
+)
+
+
```
extra line

excavador commented:
@xingao267 can you please point me to specific example of the test, where tag checked?
I re-read all rules_docker completely and did not understand what the right way to do it.
I have only single idea: in shell script I should execute bazel run <mytarget>, than try to pull container by tag and compare digests

xingao267 commented:
@excavador So when you build the new container_push target //tests/docker:push_test_tag_file you created, it should create several outputs, one is `bazel-bin/tests/docker/push_test_tag_file.digest` and one is `bazel-bin/tests/docker/push_test_tag_file`. So what I was saying in the comment is do not use the `:push_test_tag_file.digest` output, but use `:push_test_tag_file` instead. That file should contain a pusher command that includes the tag of the container. You should make sure that tag is correct.

So try to do this on your command line and you will see:

```
bazel build //tests/docker:push_test_tag_file
cat bazel-bin/tests/docker/push_test_tag_file
```
Then in the `content` you want to validate in the `file_test` rule, you need to check that file containers a regex that has the tag in it.

excavador commented:
@xingao267 thank you so much, now got it!
I knew about the executable script file but did not imagine to use it for verification.

Tests added, rebased to latest master, seems like now I fixed everything

xingao267 commented:
Thanks! Looks good!
